Use file-based db for wallet

bonomat opened the pull request:
This will improve performance because
- we don't have to generate the addresses on every restart, we can just read them from the db
- we don't have to generate 1000 addresses after are restart to miss unspent addresses: if the wallet db grows naturally, i.e. all addresses are cached, we don't have to pre-cache addresses
- we use less memory as we don't have to store the whole db in memory: particularly when having 1000s spent addresses this is a significant amount of data
- sync is faster because we don't have re-fetch the whole raw tx which is stored in the db

Note: if you have used the daemon for some time in the past the first restart can be scary because it will not immediately fetch all addresses. It will take some time until the wallet has caught up because we only sync for 100 addresses at a time.

Fixes #2158 

Note 2: this depends on a bdk fork "https://github.com/coblox/bdk", branch = "sqlite-version-bumps"  because of incompatible dependency versions. I'll see to contribute this upstream

bonomat committed:
Don't cache 1000 addresses

This unnecessary slows down syncs in most cases because all 1000 addresses need to be synced on every sync interval.
The wallet itself checks that atleast 100 addresses are cached when sync is called. This should be plenty.

bonomat committed:
Use Sled db for wallet

This will improve performance because
- we don't have to generate the addresses on every restart, we can just read them from the db
- we don't have to generate 1000 addresses after are restart to miss unspent addresses: if the wallet db grows naturally, i.e. all addresses are cached, we don't have to pre-cache addresses
- we use less memory as we don't have to store the whole db in memory: particularly when having 1000s spent addresses this is a significant amount of data
- sync is faster because we don't have re-fetch the whole raw tx which is stored in the db

Signed-off-by: Philipp Hoenisch <philipp@hoenisch.at>

da-kami commented on Cargo.toml:
```diff
@@ -26,3 +26,4 @@ maia = { git = "https://github.com/comit-network/maia", rev = "a963084189e1598ca
 maia-core = { git = "https://github.com/comit-network/maia", rev = "e419cb2cf19bed6ec53eaa3672408a6205a9b705", package = "maia-core" } # Unreleased version, pinned before maia's breaking change
 xtra_productivity = { git = "https://github.com/comit-network/xtra-productivity" } # Unreleased
 electrum-client = { git = "https://github.com/da-kami/rust-electrum-client/", branch = "fix/ignore-empty-lines" }
+bdk = { git = "https://github.com/coblox/bdk", branch = "sqlite-version-bumps" }
```
Why not a fork on itchysats?

bonomat replied:
No particular reason and we had a fork on coblox already.

da-kami reviewed (approved):
Can we test it before merging, i.e. take the testnet seed start for first time expect the balance to be `0`. Trigger sync until the balance is picked up. Then restart and the balance should be picked up without triggering another sync (because the addresses are cached in the db).

bonomat committed:
Make Wallet generic

This allows us to use an in-memory wallet for tests which gives us a better test performance.

Signed-off-by: Philipp Hoenisch <philipp@hoenisch.at>

klochowicz reviewed (approved):
looks good, I'm keen to try it out! 👍

da-kami commented on maker/src/main.rs:
```diff
@@ -65,7 +65,14 @@ async fn main() -> Result<()> {
 
-    let (wallet, wallet_feed_receiver) = wallet::Actor::new(opts.network.electrum(), ext_priv_key)?;
+    let mut wallet_dir = data_dir.clone();
+    wallet_dir.push("maker-wallet");
```
🙃 nit: We only use it twice but maybe we still want to extract it into a variable to avoid errors in the future.
Same for taker main

da-kami reviewed (approved):
Happy to give it a try. Worst case we should be able to go back to the in-memory code without any migration problems.
